src/palindrom-network-channel.js

```diff
@@ -0,0 +1,365 @@
+     * @throws {Error} network error if occured
+     */
+    async getPatchUsingHTTP(href) {
+        // we don't need to try catch here because we want the error to be thrown at whoever calls getPatchUsingHTTP
+        const data = await this._xhr(
+            href,
+            'application/json-patch+json',
+            null,
+            true
+        );
+        this.onReceive(data, href, 'GET');
+        return data;
+    }
+
+    _setRemoteUrl(remoteUrl) {
+        if (
+            this.remoteUrlSet &&
+            this.remoteUrl &&
+            this.remoteUrl != remoteUrl
+        ) {
+            const message = [
+                'Session lost.',
+                'Server replied with a different session ID than the already set one.',
+                'Possibly a server restart happened while you were working.',
+                'Please reload the page.',
+                'Previous session ID: ' + this.remoteUrl,
+                'New session ID: ' + remoteUrl
+            ].join('\n');
+
+            throw new PalindromError(message);
+        }
+        this.remoteUrlSet = true;
+        this.remoteUrl = new URL(remoteUrl, this.remoteUrl.href);
+    }
+
+    _handleLocationHeader(res) {
+        const location =
+            res.headers &&
+            (res.headers.get('x-location') || res.headers.get('location'));
+        if (location) {
+            this._setRemoteUrl(location);
+        }
+    }
+    /**
+     * Handles unsecessful HTTP requests
+     * @param error
+     */
+    async _handleFailureResponse(url, method, error) {
+        // no sufficient error information, we need to create on our own
+        var statusCode = -1;
+        var statusText = `An unknown network error has occurred. Raw message: ${
+            error.message
+        }`;
+        var reason = 'Maybe you lost connection with the server';
+        // log it for verbosity
+        console.error(error);
+
+        const message = [
+            statusText,
+            'statusCode: ' + statusCode,
+            'reason: ' + reason,
+            'url: ' + url,
+            'HTTP method: ' + method
+        ].join('\n');
+
+        this.onConnectionError(
+            new PalindromConnectionError(message, CLIENT, url, method)
+        );
+    }
+
+    /**
+     * Internal method to perform XMLHttpRequest
+     * @param url (Optional) URL to send the request. If empty string, undefined or null given - the request will be sent to window location
+     * @param accept (Optional) HTTP accept header
+     * @param data (Optional) Data payload
+     * @returns {XMLHttpRequest} performed XHR
+     */
+    async _xhr(url, accept, data, setReferer) {
+        const method = data ? 'PATCH' : 'GET';
+        const config = { headers: {}, method, credentials: 'include' };
+        const headers = config.headers;
+
+        if (data) {
+            headers['Content-Type'] = 'application/json-patch+json';
+            config.body = JSON.stringify(data);
```

Comment:
`data` should already be a string at this point, because line 187 in `palindrom.js` does it.

Reply:
This seems to be not addressed.